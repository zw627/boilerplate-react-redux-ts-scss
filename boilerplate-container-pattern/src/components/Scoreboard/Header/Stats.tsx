import React from "react";

import { StatsProps } from "Containers/Scoreboard/Header/Stats";

const Stats: React.FC<StatsProps> = ({
  playerCount,
  scoreCount,
}: StatsProps) => {
  return (
    <table className="stats">
      <tbody>
        <tr>
          <td>Players:</td>
          <td>{playerCount}</td>
        </tr>
        <tr>
          <td>Total Points:</td>
          <td>{scoreCount}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default Stats;
