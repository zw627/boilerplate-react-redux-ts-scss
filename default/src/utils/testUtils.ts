// Helper functions for Jest and Enzyme

import { ReactWrapper, ShallowWrapper } from "enzyme";
import { createStore, Store } from "redux";

import { rootReducer } from "Store/index";

// Type for parameters of setupStore()
type setupStorePars = {
  playerList?: {
    id: string;
    name: string;
    score: number;
    created: string;
    updated: string;
    isSelected: boolean;
  }[];
  selectedPlayerId?: string;
  isRunning?: boolean;
  previousTime?: number;
  elapsedTime?: number;
  isLightMode?: boolean;
  visibility?: boolean;
};

// Setup a store with some fake states for testing
export function setupStore({
  playerList = [
    {
      id: "1654A",
      name: "Alpha",
      score: 10,
      created: "11/21/2016",
      updated: "06/12/2018",
      isSelected: false,
    },
    {
      id: "1654B",
      name: "Beta",
      score: 4,
      created: "11/8/2016",
      updated: "11/9/2017",
      isSelected: false,
    },
  ],
  isRunning = false,
  previousTime = 0,
  elapsedTime = 0,
  isLightMode = false,
  visibility = false,
}: setupStorePars = {}): Store {
  return createStore(rootReducer, {
    notification: {
      storage: {
        visibility,
      },
    },
    scoreboard: {
      player: {
        playerList,
      },
      stopwatch: {
        isRunning,
        previousTime,
        elapsedTime,
      },
      theme: {
        isLightMode,
      },
    },
  });
}

// Multiple toEqual, useful when testing a table
export function multiToEqual([...args]: Array<string[] | number[]>): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    expect(arg[0]).toEqual(arg[1]);
  });
}

export function matchLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [element, number]: [string, number]
): void {
  expect(wrapper.find(element).length).toEqual(number);
}

// Multiple matchLength
export function matchMultiLength(
  wrapper: ReactWrapper | ShallowWrapper,
  [...args]: [string, number][]
): void {
  // [[a, b], [a, b], [a, b]]
  [...args].map((arg) => {
    // [a, b] (expect a equals to b)
    matchLength(wrapper, [arg[0], arg[1]]);
  });
}

// export const matchText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [element, index]: [string, number],
//   text: string
// ): void => {
//   expect(
//     wrapper
//       .find(element)
//       .at(index)
//       .text()
//   ).toEqual(text);
// };

// export const matchMultiText = (
//   wrapper: ReactWrapper | ShallowWrapper,
//   [...args]
// ): void => {
//   if ([...args].length > 0)
//     [...args].map(arg => {
//       matchText(wrapper, [arg[0], arg[1]], arg[2]);
//     });
// };
