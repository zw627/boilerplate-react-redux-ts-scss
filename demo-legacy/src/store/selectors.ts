// Selector functions (using Reselect) for Redux

import { createSelector } from "reselect";

import { AppState } from "Store/index";

// ------------------------------------------------------ //
// Notification/Storage
const notificationStorageState = (
  state: AppState
): AppState["notification"]["storage"] => state.notification.storage;

export const notiStorageVisibilitySelector = createSelector(
  notificationStorageState,
  (notificationStorage) => notificationStorage.visibility
);

// ------------------------------------------------------ //
// Scoreboard/Player

const playerListState = (
  state: AppState
): AppState["scoreboard"]["player"]["playerList"] =>
  state.scoreboard.player.playerList;

export const playerListSelector = createSelector(
  playerListState,
  (playerList) => playerList
);

export const scoreListSelector = createSelector(playerListState, (playerList) =>
  playerList.map((player) => player.score)
);

export const scoreCountSelector = createSelector(
  playerListState,
  (playerList) => playerList.reduce((total, player) => total + player.score, 0)
);

export const playerCountSelector = createSelector(
  playerListState,
  (playerList) => playerList.length
);

export const playerDetailSelector = createSelector(
  playerListState,
  (playerList) => playerList.filter((player) => player.isSelected === true)[0]
);

// ------------------------------------------------------ //
// Scoreboard/Stopwatch

const stopwatchState = (state: AppState): AppState["scoreboard"]["stopwatch"] =>
  state.scoreboard.stopwatch;

export const isRunningSelector = createSelector(
  stopwatchState,
  (stopwatch) => stopwatch.isRunning
);

export const elaspedTimeSelector = createSelector(
  stopwatchState,
  (stopwatch) => stopwatch.elapsedTime
);

// ------------------------------------------------------ //
// Scoreboard/Theme

const themeState = (state: AppState): AppState["theme"] => state.theme;

export const isLightModeSelector = createSelector(
  themeState,
  (theme) => theme.isLightMode
);
