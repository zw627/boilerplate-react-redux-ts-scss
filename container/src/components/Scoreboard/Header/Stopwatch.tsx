import React, { useEffect } from "react";

import { StopwatchProps } from "Containers/Scoreboard/Header/Stopwatch";
import formatTime from "Utils/formatTime";

const Stopwatch: React.FC<StopwatchProps> = ({
  isRunning,
  elapsedTime,
  tick,
  toggle,
  reset,
}: StopwatchProps) => {
  useEffect(() => {
    let intervalId: number | undefined;
    if (isRunning) intervalId = window.setInterval(() => tick(Date.now()), 1);
    else clearInterval(intervalId);
    return (): void => clearInterval(intervalId);
  });

  // Handlers
  function handleToggle(): void {
    toggle(Date.now());
  }
  function hanldeReset(): void {
    reset(Date.now());
  }

  return (
    <div className="stopwatch">
      <h2>Stopwatch</h2>
      <div className="stopwatch-time">{formatTime(elapsedTime)}</div>
      <button onClick={handleToggle}>{isRunning ? "Stop" : "Start"}</button>
      <button onClick={hanldeReset}>Reset</button>
    </div>
  );
};

export default Stopwatch;
