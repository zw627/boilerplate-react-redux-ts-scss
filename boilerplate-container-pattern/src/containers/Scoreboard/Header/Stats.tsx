import { connect } from "react-redux";

// Component
import Stats from "Components/Scoreboard/Header/Stats";

// Store
import { playerCountSelector, scoreCountSelector } from "Utils/selectors";
import { AppState } from "Store/index";

export interface StatsProps {
  playerCount: number;
  scoreCount: number;
}

const mapStateToProps = (state: AppState): StatsProps => ({
  playerCount: playerCountSelector(state),
  scoreCount: scoreCountSelector(state),
});

export default connect(mapStateToProps)(Stats);
