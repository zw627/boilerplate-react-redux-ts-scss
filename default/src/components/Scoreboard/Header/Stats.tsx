import React from "react";
import { useSelector } from "react-redux";

import { AppState } from "Store/index";
import { scoreCountSelector, playerCountSelector } from "Utils/selectors";

const Stats: React.FC<{}> = () => {
  // State (redux)
  const scoreCount = useSelector((state: AppState) =>
    scoreCountSelector(state)
  );
  const playerCount = useSelector((state: AppState) =>
    playerCountSelector(state)
  );

  return (
    <table className="stats">
      <tbody>
        <tr>
          <td>Players:</td>
          <td>{playerCount}</td>
        </tr>
        <tr>
          <td>Total Points:</td>
          <td>{scoreCount}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default Stats;
