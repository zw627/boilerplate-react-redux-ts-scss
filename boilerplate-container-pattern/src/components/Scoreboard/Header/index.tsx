import React from "react";

import Stats from "Containers/Scoreboard/Header/Stats";
import Stopwatch from "Containers/Scoreboard/Header/Stopwatch";

const Header: React.FC = () => {
  return (
    <div className="header">
      <Stats />
      <h1>Scoreboard</h1>
      <Stopwatch />
    </div>
  );
};

export default Header;
