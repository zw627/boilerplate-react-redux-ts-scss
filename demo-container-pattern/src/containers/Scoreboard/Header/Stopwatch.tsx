import { connect } from "react-redux";

// Component
import Stopwatch from "Components/Scoreboard/Header/Stopwatch";

// Store
import { AppState } from "Store/index";
import { isRunningSelector, elaspedTimeSelector } from "Store/selectors";
import { tick, toggle, reset } from "Store/scoreboard/stopwatch/actions";

interface StopwatchStateProps {
  isRunning: boolean;
  elapsedTime: number;
}

export interface StopwatchProps extends StopwatchStateProps {
  tick: typeof tick;
  toggle: typeof toggle;
  reset: typeof reset;
}

const mapStateToProps = (state: AppState): StopwatchStateProps => ({
  isRunning: isRunningSelector(state),
  elapsedTime: elaspedTimeSelector(state),
});

const actionCreators = {
  tick,
  toggle,
  reset,
};

export default connect(mapStateToProps, actionCreators)(Stopwatch);
