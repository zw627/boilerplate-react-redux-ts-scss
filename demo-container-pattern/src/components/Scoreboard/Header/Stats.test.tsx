import React from "react";
import { Provider } from "react-redux";
import { mount, ReactWrapper } from "enzyme";

import Stats from "Components/Scoreboard/Header/Stats";
import { setupStore, multiToEqual } from "Utils/testUtils";
import { playerCountSelector, scoreCountSelector } from "Store/selectors";

import { AppState } from "Store/index";
import { StatsProps } from "Containers/Scoreboard/Header/Stats";

// Pass down the props from mock store to wrapper
function setupProps(state = setupStore().getState() as AppState): StatsProps {
  return {
    playerCount: playerCountSelector(state),
    scoreCount: scoreCountSelector(state),
  };
}

function setupWrapper(
  mockStore = setupStore(),
  props = setupProps(mockStore.getState())
): ReactWrapper {
  return mount(
    <Provider store={mockStore}>
      <Stats {...props} />
    </Provider>
  ) as ReactWrapper;
}

describe("Scoreboard/Stats", () => {
  let wrapper: ReactWrapper;

  // Setup wrapper
  beforeEach(() => {
    wrapper = setupWrapper();
  });

  test("compare to the last snapshot", () => {
    expect(wrapper.find(Stats)).toMatchSnapshot();
  });

  /* eslint jest/expect-expect: ["error", { "assertFunctionNames": ["expect", "multiToEqual"] }] */
  it("should display player and point count", () => {
    function mapRowColumns(index: number): string[] {
      return wrapper
        .find("tr")
        .at(index)
        .find("td")
        .map((column) => column.text());
    }
    const firstRowColumns = mapRowColumns(0);
    const secondRowColumns = mapRowColumns(1);
    multiToEqual([
      [firstRowColumns.length, 2],
      [firstRowColumns[0], "Players:"],
      [firstRowColumns[1], "2"],
      [secondRowColumns.length, 2],
      [secondRowColumns[0], "Total Points:"],
      [secondRowColumns[1], "14"],
    ]);
  });
});
