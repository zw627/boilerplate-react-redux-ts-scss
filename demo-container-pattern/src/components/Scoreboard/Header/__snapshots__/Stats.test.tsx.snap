// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scoreboard/Stats compare to the last snapshot 1`] = `
<Stats
  playerCount={2}
  scoreCount={14}
>
  <table
    className="stats"
  >
    <tbody>
      <tr>
        <td>
          Players:
        </td>
        <td>
          2
        </td>
      </tr>
      <tr>
        <td>
          Total Points:
        </td>
        <td>
          14
        </td>
      </tr>
    </tbody>
  </table>
</Stats>
`;
